docs/SPECS.md

```diff
@@ -0,0 +1,770 @@
+[BLS12-381](https://hackmd.io/@benjaminion/bls12-381). All points on the curve are sent using the compressed form.
+
+The implementation that drand uses is located in the
+[bls12-381](https://github.com/drand/bls12-381) repo.
+**Hash to curve**:The hash-to-curve algorithm is derived from the [RFC
+v7](https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-07). 
+
+### Distributed Public Key
+
+The distributed public key is a list of BLS12-381 G1 points that represents the
+public polynomial created during the DKG protocol.
+```go
+// DistPublic represents the distributed public key generated during a DKG. This
+// is the information that can be safely exported to end users verifying a
+// drand signature. It is the list of all commitments of the coefficients of the
+// private distributed polynomial.
+type DistPublic struct {
+    // points on the BLS12-381 G1 curve
+	Coefficients [][]byte 
+}
+```
+The coefficients are needed to verify partial beacon signatures. 
+The first coefficient is needed to verify the final beacon signature.
+
+### Beacon Signature
+
+A beacon signature is a regular [BLS signature](https://www.iacr.org/archive/asiacrypt2001/22480516.pdf) over the message:
+```go
+func Message(currRound uint64, prevSig []byte) []byte {
+	h := sha256.New()
+	h.Write(prevSig)
+	h.Write(roundToBytes(currRound))
+	return h.Sum(nil)
+}
+```
+The ciphersuite used is:
+```go
+var Domain = []byte("BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_NUL_")
+```
+
+It can be verified using the first coefficient of the distributed public key
+stored in the group configuration:
+
+### Partial Beacon Signature
+
+A partial beacon signature is created over the same input as the beacon
+signature. However, the node's index is prefixed on the first two bytes to the
+signature.
+```go
+func concatenate(signature []byte, index uint16) []byte {
+    var buffer bytes.Buffer
+    binary.Write(buffer, binary.BigEndian, index)
+    buffer.Write(sig)
+    return buffer.Bytes()
+}
+```
+
+**Verifying**: The public key to validate a partial signature is derived from
+the distributed public key. Namely, the public key is the evaluation of the
+polynomial whose coefficients are the points of the distributed key, at the
+index of the signer + 1. We need to increment the signer's index since the
+evaluation point 0 corresponds to the public key but no nodes posseses the
+private key. More on that in the DKG sections below.
+To evaluate the polynomial, one can use the following routine:
+```go
+// i is the index of the signer
+// commits are the points of the distributed public key
+func Eval(i int,commits []Point) Point {
+	xi := Scalar().SetInt64(1 + int64(i)) // x-coordinate of this share
+	v := Point().Null()
+	for j := p.Threshold() - 1; j >= 0; j-- {
+		v.Mul(xi, v)
+		v.Add(v, commits[j])
+	}
+	return v
+}
+```
+
+## External API
+
+## Control API
+
+## THINGS TO REVIEW
+
+* Setup phase: now it doesn't require any manual downloading from operators, and
```

Comment:
My imagination of what might be easiest is:
* each participant can at their leisure run `drand identity generate` or similar to create their local key, and produce a document with their public address and public key.
* each participant gives that proposed identity to a coordinator, who generates a manifest document, containing the list of all groups (or everyone passes them around to everyone else; just everyone needs to have a single document with the concatenation of all the addresses and keys as the input.
* Note; the previous 2 steps can happen out of band and on a different computer from the actual deployed node.
* drand is then provisioned and spun up by each participant. it is given the manifest and the local node's private key.
* nodes attempt to contact all other nodes in the manifest on startup and periodically thereafter. The node with the lowest hash is implicitly the leader. Once that node successfully connects to all other nodes it begins the setup phase automatically.

cc @hsanjuan

Reply:
So the first 4 points is what drand was just a few weeks ago ;) 
It turned out it was a real hassle to distribute the group config file, ppl sometimes got the wrong one, and then ppl accidently deleted their private key so a new group file must be generated again and passed down to the operators again,etc.
The goal here was to reduce the amount of manual work operators needed to do. Now they only need to run one command without downloading or pushing anything and i've received good feedback about it.

I would be fine to go back to a more manual approach but we should try to minimize the amount of manual work operators need to do to setup anything. The ideas mentionned here and in the issue of easing up the setup process by having drand automatically detects a change and not needing to issue a command is a good step forward.

Reply:
it sounds like longer term, you might be wanting to bootstrap with a PAKE (e.g. https://github.com/agl/panda/blob/master/panda.go) - e.g. the nodes all put group membership announcements under the shared secret and all end up with active addresses of each other on which to begin the interactive phase of setup.

Shorter term, The reason to start with "manifest of participant keys and addresses" is that it means deployment from there can be fully automated.

We can build a quick web app style thing for the group members to all upload their keys on day/hour 0 and then drand could ingest from the url and begin doing its thing at day/hour 1. we could maybe have a second manual 'go' step where everyone looks at the manifest and sanity checks it before they join, but it could also 'just happen' and people are responsible for looking at the group their node shows once they're up and running.

Reply:
The current `drand share` is a brittle process now:

* Must ensure full connectivity outgoing and incoming
* Must make sure generate-keypair has been run with the correct hostname
* Must ensure everyone does it at the same time
* The `--leader` part of the command blocks but it can also be cancelled and the sharing process continues in the background, but then it can be run again and I have no idea what happens (does it restart? Does it continue? Do I need to restart the drand node?)
* I have seen it fail with really weird errors and then work after a couple of retries and restart of the nodes.
* It is very difficult to build anything automatically deployable with this
* The whole process need to be done in sync by everyone. It if fails, the whole "let's relaunch this thing needs to be done in sync by everyone.

To have a generator which takes a number of public keys and creates the group.toml file seems a much better approach. The configuration can be deployed following standard methods, not everyone needs to do it at the same time, connectivity problems can be debugged separately without affecting the process for everyone else, there is no need to elect a `leader` and have a special role...